File: day13_part2_2017.php

<?php

$file = fopen("input.txt", "r");
if ($file === false) {
    echo "Error reading file\n";
    return;
}

$firewall = [];
while (($line = fgets($file)) !== false) {
    $fields = explode(": ", $line);
    $depth = (int)$fields[0];
    $range = (int)$fields[1];
    $firewall[$depth] = ["Range" => $range, "Position" => 0, "Direction" => 1];
}

$delay = 0;
while (true) {
    if (passThrough($firewall, $delay)) {
        break;
    }
    $delay++;
}

echo $delay . "\n";

function passThrough($firewall, $delay) {
    foreach ($firewall as $depth => $scanner) {
        if (($depth + $delay) % (2 * ($scanner["Range"] - 1)) == 0) {
            return false;
        }
    }
    return true;
}
?>

File: objectivec/day13_part2_2017.m

#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSError *error;
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        
        NSMutableDictionary *firewall = [NSMutableDictionary dictionary];
        for (NSString *line in lines) {
            NSArray *components = [line componentsSeparatedByString:@": "];
            firewall[components[0]] = components[1];
        }
        
        int severity = 0;
        for (NSString *depth in firewall.allKeys) {
            int range = [firewall[depth] intValue];
            if ([depth intValue] % ((range - 1) * 2) == 0) {
                severity += [depth intValue] * range;
            }
        }
        
        printf("Part 1: %d\n", severity);
        
        int delay = 0;
        while (true) {
            BOOL caught = NO;
            for (NSString *depth in firewall.allKeys) {
                int range = [firewall[depth] intValue];
                if (([depth intValue] + delay) % ((range - 1) * 2) == 0) {
                    caught = YES;
                    break;
                }
            }
            if (!caught) {
                break;
            }
            delay++;
        }
        
        printf("Part 2: %d\n", delay);
    }
    return 0;
}
